src/main/scala/reporting/Converter.scala

```diff
@@ -0,0 +1,503 @@
+        val psfVal = evaluateTerm(psf, model).asValueEntry
+        val arg = Seq(psfVal, snapVal)
+        val result = getFunctionValue(model, lookupFuncName, arg, sorts.Snap)
+        result
+
+      case _ =>
+        OtherEntry(s"$term", "unhandled")
+    }
+  }
+
+  def extractHeap(h: Iterable[Chunk], model: Model): ExtractedHeap = {
+    var entries: Vector[HeapEntry] = Vector()
+    h foreach {
+      case c @ BasicChunk(FieldID, _, _, _, _) =>
+        val entry = extractField(c, model)
+        entries = entries :+ entry
+      case c @ BasicChunk(PredicateID, _, _, _, _) =>
+        val entry = extractPredicate(c, model)
+        entries = entries :+ entry
+      case c: BasicChunk =>
+        entries = entries :+ UnresolvedHeapEntry(c, "Magic Wands not supported")
+      case c =>
+        entries =
+          entries :+ UnresolvedHeapEntry(c, "Non-basic chunks not supported")
+    }
+    ExtractedHeap(entries)
+  }
+
+  def extractField(chunk: BasicChunk, model: Model): HeapEntry = {
+    val fieldname = chunk.id.name
+    val recv = evaluateTerm(chunk.args.head, model)
+    val recvVar: Try[VarEntry] = Try(recv.asInstanceOf[VarEntry])
+    // this should always be of type VarEntry
+
+    val perm: Option[Rational] = chunk.perm match {
+      case p: PermLiteral => Some(p.literal)
+      //if there are other types of values for permissions they should be added here
+      case _ => None
+    }
+    val value = evaluateTerm(chunk.snap, model)
+    recvVar match {
+      case Success(varEntry) =>
+        FieldHeapEntry(varEntry, fieldname, perm, chunk.snap.sort, value)
+      case _ =>
+        UnresolvedHeapEntry(
+          chunk,
+          "the receiver term is not a variable (unexpected)"
+        )
+    }
+  }
+
+  def extractPredicate(chunk: BasicChunk, model: Model): HeapEntry = {
+    //this might be too simple for some cases but for prusti to tell if some
+    //variable is part of a class it should be good enough
+    //not really sure if the snap value should be added, seems to be same as one of
+    //the args in most cases.
+    val argsEval = chunk.args.map(x => evaluateTerm(x, model))
+    val entry = PredHeapEntry(chunk.id.toString, argsEval)
+    entry
+  }
+
+  def extractSequence(
+      model: Model,
+      heap: ExtractedHeap,
+      nullRefName: String,
+      name: String,
+      elementSort: Sort,
+      encountered: Set[String]
+  ): ExtractedModelEntry = {
+    if (!encountered.contains(name)) {
+      val newEncountered = encountered + name
+      //hopefully Seq_len can only contain integers
+      val lenTry: Try[Int] = Try(
+        getFunctionValue(
+          model,
+          "Seq_length",
+          Seq(ConstantEntry(name)),
+          sorts.Int
+        ).asInstanceOf[LitIntEntry].value.toInt
+      )
+      lenTry match {
+        case Success(x) =>
+          val len: Int = x
+          var values: Vector[ExtractedModelEntry] = Vector()
+          for (i <- 0 to len) {
```

Comment:
This might be the cause of the bug @Felalolf commented.
```suggestion
          for (i <- 0 until len) {
```

Reply:
Oh yes, thank you @Felalolf @Aurel300 ! I should have noticed that.